src/tunnel.cc

```diff
@@ -1307,37 +1301,72 @@ TunnelStateData::saveError(ErrorState *error)
 {
     debugs(26, 4, savedError << " ? " << error);
     assert(error);
+    // nobody saves finalError (or makes finalError equal to savedError)
+    assert(!finalError || (finalError != error && finalError != savedError));
+    if (savedError == error)
+        return;
     delete savedError; // may be nil
     savedError = error;
 }
 
-/// Starts sending the given error message to the client, leading to the
-/// eventual transaction termination. Call with savedError to send savedError.
+/// Makes sure the transaction is terminated, immediately (if possible) or after
+/// reporting the previously saved error to the client and/or finishing in-progress
+/// I/Os (as needed).
+/// All callers guarantee savedError existence, which becomes the final error during
+/// the first call. Safe to call multiple times.
 void
-TunnelStateData::sendError(ErrorState *finalError, const char *reason)
+TunnelStateData::terminateWithSavedError(const char *reason)
 {
     debugs(26, 3, "aborting transaction for " << reason);
+    assert(savedError);
 
-    if (request)
-        request->hier.stopPeerClock(false);
+    if (finalError) {
+        debugs(26, 4, "the final error already exists: " << finalError << ", discarding new error: " << savedError);
+        assert(finalError != savedError);
+        return;
+    }
 
-    if (opening())
-        cancelOpening(reason);
+    finalError = savedError;
+    savedError = nullptr;
 
-    assert(finalError);
+    // TODO: if we are still writing to the peer, we need to do this later,
+    // when all data is written (or the server connection is closed).
```

Comment:
@rousskov , if you manage to come up with a better explanation, I am ready to adjust it accordingly.